pkg/pgmodel/ingestor/ingestor.go

```diff
@@ -52,12 +57,73 @@ func NewPgxIngestorForTests(conn pgxconn.PgxConn) (*DBIngestor, error) {
 //     tts the []Timeseries to insert
 //     req the WriteRequest backing tts. It will be added to our WriteRequest
 //         pool when it is no longer needed.
-func (ingestor *DBIngestor) Ingest(r *prompb.WriteRequest) (uint64, error) {
-	var (
-		err       error
-		totalRows uint64
+func (ingestor *DBIngestor) Ingest(r *prompb.WriteRequest) (numSamples uint64, numMetadata uint64, err error) {
+	defer func() {
+		// WriteRequests can contain pointers into the original buffer we deserialized
+		// them out of, and can be quite large in and of themselves. In order to prevent
+		// memory blowup, and to allow faster deserializing, we recycle the WriteRequest
+		// here, allowing it to be either garbage collected or reused for a new request.
+		// In order for this to work correctly, any data we wish to keep using (e.g.
+		// samples) must no longer be reachable from req.
+		FinishWriteRequest(r)
+	}()
```

Comment:
I'm not sure about this... I understand why you did it like this but we are not freeing the memory as soon as its not used like before. 

I believe there is a better way to support this by checking for samples and/or metadata. If either is missing, you can flush the write request. If both are present, just remove the samples or metadata from the request.